WebCryptoAPI/util/helpers.js

```diff
@@ -0,0 +1,240 @@
+}
+
+// Is key a CryptoKey object with correct algorithm, extractable, and usages?
+// Is it a secret, private, or public kind of key?
+function assert_goodCryptoKey(key, algorithm, extractable, usages, kind) {
+    var correctUsages = [];
+
+    var registeredAlgorithmName;
+    registeredAlgorithmNames.forEach(function(name) {
+        if (name.toUpperCase() === algorithm.name.toUpperCase()) {
+            registeredAlgorithmName = name;
+        }
+    });
+
+    assert_equals(key.constructor, CryptoKey, "Is a CryptoKey");
+    assert_equals(key.type, kind, "Is a " + kind + " key");
+    if (key.type === "public") {
+        extractable = true; // public keys are always extractable
+    }
+    assert_equals(key.extractable, extractable, "Extractability is correct");
+
+    assert_equals(key.algorithm.name, registeredAlgorithmName, "Correct algorithm name");
+    assert_equals(key.algorithm.length, algorithm.length, "Correct length");
+    if (["HMAC", "RSASSA-PKCS1-v1_5", "RSA-PSS"].includes(registeredAlgorithmName)) {
+        assert_equals(key.algorithm.hash.name.toUpperCase(), algorithm.hash.toUpperCase(), "Correct hash function");
+    }
+
+    // usages is expected to be provided for a key pair, but we are checking
+    // only a single key. The publicKey and privateKey portions of a key pair
+    // recognize only some of the usages appropriate for a key pair.
+    if (key.type === "public") {
+        ["encrypt", "verify", "wrapKey"].forEach(function(usage) {
+            if (usages.includes(usage)) {
+                correctUsages.push(usage);
+            }
+        });
+    } else if (key.type === "private") {
+        ["decrypt", "sign", "unwrapKey", "deriveKey", "deriveBits"].forEach(function(usage) {
+            if (usages.includes(usage)) {
+                correctUsages.push(usage);
+            }
+        });
+    } else {
+        correctUsages = usages;
+    }
+
+    assert_equals((typeof key.usages), "object", key.type + " key.usages is an object");
+    assert_not_equals(key.usages, null, key.type + " key.usages isn't null");
+
+    // The usages parameter could have repeats, but the usages
+    // property of the result should not.
+    var usageCount = 0;
+    key.usages.forEach(function(usage) {
+        usageCount += 1;
+        assert_in_array(usage, correctUsages, "Has " + usage + " usage");
+    });
+    assert_equals(key.usages.length, usageCount, "usages property is correct");
+}
+
+
+// The algorithm parameter is an object with a name and other
+// properties. Given the name, generate all valid parameters.
+//
+// You can have a "short run" for RSA by including a second, true
+// parameter. Useful if you're testing one variant of the
+// algorithm name exhaustively, so need less exhaustive tests
+// for other variants.
+function allAlgorithmSpecifiersFor(algorithmName, shortRun) {
+    var results = [];
+    var hashes = ["SHA-1", "SHA-256", "SHA-384", "SHA-512"];
+    var keyLengths = [1024, 2048, 3072, 4096];
```

Comment:
I'm going to make all of the changes you suggest here for RSA. I tried to
cover a very broad range of parameters, but the reality is if they cover a
single combination of modulus length/public exponent then they properly
support RSA per the spec. So I'll probably cut it back to just 2048/0x10001.

For AES, the spec does list key sizes to support, and the cost in time of
testing them all is not high, so I'm inclined to keep them all in the tests.

Charlie

On Mon, May 23, 2016 at 10:02 AM, bifurcation <notifications@github.com>
wrote:

> In WebCryptoAPI/util/helpers.js
> <https://github.com/w3c/web-platform-tests/pull/3016#discussion_r64227562>
> :
>
> > +    });
> > +    assert_equals(key.usages.length, usageCount, "usages property is correct");
> > +}
> > +
> > +
> > +// The algorithm parameter is an object with a name and other
> > +// properties. Given the name, generate all valid parameters.
> > +//
> > +// You can have a "short run" for RSA by including a second, true
> > +// parameter. Useful if you're testing one variant of the
> > +// algorithm name exhaustively, so need less exhaustive tests
> > +// for other variants.
> > +function allAlgorithmSpecifiersFor(algorithmName, shortRun) {
> > +    var results = [];
> > +    var hashes = ["SHA-1", "SHA-256", "SHA-384", "SHA-512"];
> > +    var keyLengths = [1024, 2048, 3072, 4096];
>
> In general, the combinatorics are overly aggressive here. It seems like
> the goal of these tests should be to test the *API*, vs. the underlying
> crypto library. It seems like there are three classes of parameter here:
>
>    1. Things that get passed straight through to the library: AES and RSA
>    key sizes, HMAC lengths
>    2. Things that are algorithmically transformed: RSA public moduli
>    3. Things that need to be looked up in a table: Hash and curve names
>
> For (1) and (2), it suffices to test one value to verify that the
> pass-through / algorithmic transform behaves correctly. Only for (3) do you
> need to test all the values.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/3016/files/5d5c86c779c6218dd1f48847b0c952f440fdc22e#r64227562>
>